src/windows/win_broker_os_interface.cpp

```diff
@@ -0,0 +1,227 @@
+ ******************************************************************************
+ * This file is a part of RDMnetBroker. For more information, go to:
+ * https://github.com/ETCLabs/RDMnetBroker
+ *****************************************************************************/
+
+#include "win_broker_os_interface.h"
+
+#include <iostream>
+#include <memory>
+#include <cassert>
+#include <Windows.h>
+#include <ShlObj.h>
+#include <datetimeapi.h>
+#include "service_utils.h"
+
+constexpr const WCHAR kRelativeConfFileName[] = L"\\ETC\\RDMnetBroker\\broker.conf";
+static const std::vector<std::wstring> kRelativeLogFilePath = {L"ETC", L"RDMnetBroker"};
+static const std::wstring kLogFileName = L"broker.log";
+static constexpr int kMaxLogRotationFiles = 5;
+
+std::string ConvertWstringToUtf8(const std::wstring& str)
+{
+  // Convert the path to UTF-8
+  int size_needed = WideCharToMultiByte(CP_UTF8, 0, str.c_str(), -1, NULL, 0, NULL, NULL);
+  if (size_needed > 0)
+  {
+    auto buf = std::make_unique<char[]>(size_needed);
+    int convert_res = WideCharToMultiByte(CP_UTF8, 0, str.c_str(), -1, buf.get(), size_needed, NULL, NULL);
+    if (convert_res > 0)
+    {
+      return buf.get();
+    }
+  }
+  return std::string{};
+}
+
+WindowsBrokerOsInterface::WindowsBrokerOsInterface()
+{
+  PWSTR program_data_path;
+  HRESULT get_known_folder_res = SHGetKnownFolderPath(FOLDERID_ProgramData, 0, NULL, &program_data_path);
+  if (get_known_folder_res == S_OK)
+  {
+    program_data_path_ = program_data_path;
+    CoTaskMemFree(program_data_path);
+
+    // Build the log file path on top of it
+    log_file_path_ = program_data_path_;
+    for (const auto& intermediate_dir : kRelativeLogFilePath)
+    {
+      log_file_path_ += L"\\" + intermediate_dir;
+    }
+    log_file_path_ += L"\\" + kLogFileName;
+  }
+}
+
+WindowsBrokerOsInterface::~WindowsBrokerOsInterface()
+{
+  if (log_file_)
+    fclose(log_file_);
+}
+
+std::string WindowsBrokerOsInterface::GetLogFilePath() const
+{
+  if (log_file_path_.empty())
+    return std::string{};
+
+  return ConvertWstringToUtf8(log_file_path_);
+}
+
+bool WindowsBrokerOsInterface::OpenLogFile()
+{
+  assert(!log_file_);
+
+  if (program_data_path_.empty())
+  {
+    std::cout << "FATAL: Could not get location of ProgramData directory.\n";
+    return false;
+  }
+
+  std::wstring intermediate_dir_path = program_data_path_;
+  for (const auto& intermediate_dir : kRelativeLogFilePath)
+  {
+    intermediate_dir_path += L"\\" + intermediate_dir;
+    if (!CreateDirectoryW(intermediate_dir_path.c_str(), NULL))
+    {
+      DWORD error = GetLastError();
+      if (error != ERROR_ALREADY_EXISTS)
+      {
+        // Something went wrong creating an intermediate directory.
+        std::wcout << L"FATAL: Error creating intermediate directory \"" << intermediate_dir << L"\" in ProgramData: "
+                   << error << L"\n";
+        return false;
+      }
+    }
+  }
+
+  DWORD rotate_result = RotateLogs();
```

Comment:
That's a rotate!

Reply:
![russian_coat_check](https://user-images.githubusercontent.com/5142299/67776144-ea20c800-fa2d-11e9-8abb-3b28738a1485.jpg)